src/Microsoft.OData.Core/UriParser/Binders/MetadataBindingUtils.cs

```diff
@@ -58,19 +58,31 @@ internal static SingleValueNode ConvertToTypeIfNeeded(SingleValueNode source, IE
                 }
 
                 ConstantNode constantNode = source as ConstantNode;
-                if (constantNode != null && constantNode.Value != null && source.TypeReference.IsString() && targetTypeReference.IsEnum())
+                // Check if the source node is a constant node, not null, and the source type is either string or integral
+                // and the target type is an enum.
+                if (constantNode != null && constantNode.Value != null && (source.TypeReference.IsString() || source.TypeReference.IsIntegral()) && targetTypeReference.IsEnum())
                 {
                     string memberName = constantNode.Value.ToString();
                     IEdmEnumType enumType = targetTypeReference.Definition as IEdmEnumType;
                     if (enumType.Members.Any(m => string.Compare(m.Name, memberName, StringComparison.Ordinal) == 0))
                     {
                         string literalText = ODataUriUtils.ConvertToUriLiteral(constantNode.Value, default(ODataVersion));
-                        return new ConstantNode(new ODataEnumValue(constantNode.Value.ToString(), targetTypeReference.Definition.ToString()), literalText, targetTypeReference);
+                        return new ConstantNode(new ODataEnumValue(memberName, enumType.ToString()), literalText, targetTypeReference);
                     }
-                    else
+
+                    // If the member name is an integral value, we should try to convert it to the enum member name.
+                    if (int.TryParse(memberName, out int memberIntegralValue))

```

Comment:
What if someone has defined their enum as follows:
```csharp
public enum MyEnum : long
{
    MyEnumMember = (long)int.MaxValue + (long)int.MaxValue
}
```
Maybe we should use `long.TryParse` since it'd work for both `int` and `long` scenarios...

Reply:
You should probably add a test case for that to confirm...

Reply:
@gathogojr 

I have used `long.TryParse` as it is able to handle the following integral scenarios:
```cs
case EdmPrimitiveTypeKind.Int64:
case EdmPrimitiveTypeKind.Int32:
case EdmPrimitiveTypeKind.Int16:
case EdmPrimitiveTypeKind.Byte:
case EdmPrimitiveTypeKind.SByte:
```

I have added the following tests for the scenarios you mentioned:
- [IfTypePromotionNeeded_SourceIsLongAsStringMemberValueAndTargetIsEnum_ConstantNodeIsCreated()](https://github.com/OData/odata.net/blob/fix/2542-filter-with-enum-integer/test/FunctionalTests/Microsoft.OData.Core.Tests/UriParser/Binders/MetadataBindingUtilsTests.cs#L113)
- [IfTypePromotionNeeded_SourceIsLongMemberValueAndTargetIsEnum_ConstantNodeIsCreated()](https://github.com/OData/odata.net/blob/fix/2542-filter-with-enum-integer/test/FunctionalTests/Microsoft.OData.Core.Tests/UriParser/Binders/MetadataBindingUtilsTests.cs#L75)
- [IfTypePromotionNeeded_SourceIsHugeLongMemberValueAndTargetIsEnum_ConstantNodeIsCreated()](https://github.com/OData/odata.net/blob/fix/2542-filter-with-enum-integer/test/FunctionalTests/Microsoft.OData.Core.Tests/UriParser/Binders/MetadataBindingUtilsTests.cs#L94)
- [IfTypePromotionNeededForEnum_SourceIsIntegralMemberValueInStringAndTargetIsEnumType_ConstantNodeIsCreated()](https://github.com/OData/odata.net/blob/fix/2542-filter-with-enum-integer/test/FunctionalTests/Microsoft.OData.Core.Tests/UriParser/Binders/MetadataBindingUtilsTests.cs#L132)